src/persist-txn/src/txn_read.rs

```diff
@@ -321,19 +340,85 @@ enum WaitTs<T> {
 }
 
 #[derive(Debug)]
-struct TxnsReadTask<T: Timestamp + Lattice + Codec64, C: TxnsCodec = TxnsCodecDefault> {
+struct TxnsReadTask<T: Timestamp + Lattice + Codec64> {
     rx: mpsc::Receiver<TxnsReadCmd<T>>,
-    cache: TxnsCache<T, C>,
+    cache: TxnsCacheState<T>,
+    pending_waits: Vec<PendingWait<T>>,
+    frontier: T,
+}
+
+/// A pending "wait" notification that we will complete once the frontier
+/// advances far enough.
+#[derive(Debug)]
+struct PendingWait<T: Timestamp + Lattice + Codec64> {
+    ts: WaitTs<T>,
+    tx: Option<oneshot::Sender<()>>,
 }
 
-impl<T, C> TxnsReadTask<T, C>
+impl<T: Timestamp + Lattice + Codec64> PendingWait<T> {
+    /// Returns `true` if this [PendingWait] is completed.
+    ///
+    /// A pending wait is completed when the frontier advances far enough or
+    /// when the receiver side hangs up.
+    fn maybe_complete(&mut self, frontier: &T) -> bool {
+        if self.tx.is_none() {
+            // Already completed!
+            return true;
+        }
+
+        match &self.ts {
+            WaitTs::GreaterEqual(ts) => {
+                if frontier >= ts {
+                    let _ = self.tx.take().expect("known to exist").send(());
+                    return true;
+                }
+            }
+            WaitTs::GreaterThan(ts) => {
+                if frontier > ts {
+                    let _ = self.tx.take().expect("known to exist").send(());
+                    return true;
+                }
+            }
+        };
+
+        if let Some(tx) = self.tx.as_ref() {
+            if tx.is_closed() {
+                // Receiver dropped, so also complete.
+                self.tx.take();
+                return true;
+            }
+        }
+
+        false
+    }
+}
+
+impl<T> TxnsReadTask<T>
 where
     T: Timestamp + Lattice + TotalOrder + StepForward + Codec64,
-    C: TxnsCodec,
 {
     async fn run(&mut self) {
         while let Some(cmd) = self.rx.recv().await {
             match cmd {
+                TxnsReadCmd::Updates { entries, frontier } => {
+                    tracing::trace!(
+                        "updates from subscribe task at ({:?}): {:?}",
+                        frontier,
+                        entries
+                    );
+
+                    self.cache.push_entries(entries, frontier.clone());
+
+                    // Respond to waits and retain those that still have to
+                    // wait.
+                    self.pending_waits.retain_mut(|wait| {
```

Comment:
It's `O(n^2)` to do this on every frontier advancement, but if you keep them in a heap, you can just pop them off until one of them returns false and then stop (I carefully constructed the Ord impl such that everything after that would also return false)

Reply:
Is it `O(n^2)` because `retain_mut` does back-shifting for every removed element?

My thinking here was:
 - I have a thing that I'm trying to use simpler data structures if possible, so an array and sorting rather than a heap/queue/hashmap, say. Because those can often get you quite a long way and are cheaper for a smaller amount of elements.
 - I was assuming that pending waits normally don't "survive" the next `Updates` message.
 - With :point_up: and vec/retain, insertion is `O(1)` and for every update we do one sweep of the vec that retires almost all waits.
 - I don't need a more complicated scheme for cancelling waits.

I'm :100: willing to change it for a heap/queue or sth else if you still think that's better!

Reply:
I had assumed `retain` was `O(n)` before but didn't check. I now looked at the code of `retain_mut` and I think it's really only `O(n)`: we do two sweeps across the vec, but each element is moved only once, to fill holes, and then never touched again.

Reply:
> I was assuming that pending waits normally don't "survive" the next Updates message.

This is also what I'd expect, but when I was testing locally I want to say I was seeing waits come in with timestamps of u64::MAX. So the `O(n^2)` I was talking about was more like `O(pending_waits * progress_advancements)` more than anything to do with how retain_mut works.

I also generally prefer to keep things simple when I can but I think I'm much more stressed about the worst case behavior here than I am about shaving off the constant factor on the best/expected case

Reply:
> I'm much more stressed about the worst case behavior here than I am about shaving off the constant factor on the best/expected case

got it! I pushed a commit that aims at implementing what you sketched [in your other comment](https://github.com/MaterializeInc/materialize/pull/24364#discussion_r1449558065)